samza-kafka/src/main/scala/org/apache/samza/checkpoint/kafka/KafkaCheckpointManager.scala

```diff
@@ -187,92 +174,83 @@ class KafkaCheckpointManager(
-                return
-              }
-              KafkaUtil.maybeThrowException(errorCode)
+    val ssp: SystemStreamPartition = new SystemStreamPartition(systemName, checkpointTopic, new Partition(0))
+    val systemConsumer = getSystemConsumer()
+    val (oldestOffset, newestOffset) = getEarliestOffset(checkpointTopic, new Partition(0))
+    systemConsumer.register(ssp, oldestOffset); // checkpoint stream should always be read from the beginning
+    systemConsumer.start();
+
+    var count = 0
+    try {
+      var done = false
+      var currentOffset = -1L
+      val newestOffsetLong = newestOffset.toLong
+      while (currentOffset < newestOffsetLong) {
+        // Map[SystemStreamPartition, List[IncomingMessageEnvelope]]
+        val envelopes = systemConsumer.poll(Collections.singleton(ssp), 1000)
+
+        val messages: util.List[IncomingMessageEnvelope] = envelopes.get(ssp)
+        val messagesSize = if (messages != null) messages.size else -1
+        info("CheckpointMgr read %s envelopes (%s messages) from ssp %s"
+                     format (envelopes.size(), messagesSize, ssp))
+        if (envelopes.isEmpty || messages.isEmpty) {
+          // fully read the log
+          info("Got empty list of messages")
+          //done = true
+        }
+        else {
+          count += messages.size()
+          // check the key
+          for (msg: IncomingMessageEnvelope <- messages) {
+            val key = msg.getKey.asInstanceOf[Array[Byte]]
+            currentOffset = msg.getOffset().toLong
+            if (key == null)
+              throw new KafkaUtilException("While reading checkpoint stream encountered message without key.")
+
+            val checkpointKey = KafkaCheckpointLogKey.fromBytes(key)
+
+            if (!shouldHandleEntry(checkpointKey)) {
+              info("Skipping " + entryType + " entry with key " + checkpointKey)
             }
-
-            for (response <- fetchResponse.messageSet(checkpointTopic, 0)) {
-              offset = response.nextOffset
-              startingOffset = Some(offset) // For next time we call
-
-              if (!response.message.hasKey) {
-                throw new KafkaUtilException("Encountered message without key.")
-              }
-
-              val checkpointKey = KafkaCheckpointLogKey.fromBytes(Utils.readBytes(response.message.key))
-
-              if (!shouldHandleEntry(checkpointKey)) {
-                debug("Skipping " + entryType + " entry with key " + checkpointKey)
-              } else {
-                handleEntry(response.message.payload, checkpointKey)
-              }
+            else {
+              val bbuf: java.nio.ByteBuffer = ByteBuffer.wrap(msg.getMessage.asInstanceOf[Array[Byte]])
```

Comment:
Prefer variables to be named what they actually mean semantically (instead of what their types are):
s/bbuf/checkpointPayLoad

Reply:
Maybe checkpointMessage? Or 'checkpointPayload'. Don't need type annotation.

Reply:
I was so surprised that handle() method takes ByteBuffer instead of byte[], that I even named the variable after it.
Done.